tower-log/src/lib.rs

```diff
@@ -0,0 +1,452 @@
+    };
+    (level: $level:expr, $inner:expr) => {
+        $crate::LogResponses::new($inner)
+            .in_module(module_path!())
+            .at_location(file!(), line!())
+            .at_level($level)
+    };
+}
+/// Logs error responses.
+#[derive(Clone, Debug)]
+pub struct LogErrors<T> {
+    inner: T,
+    level: log::Level,
+    target: Option<&'static str>,
+    module_path: Option<&'static str>,
+    file: Option<&'static str>,
+    line: Option<u32>,
+}
+
+/// Logs successful responses.
+#[derive(Clone, Debug)]
+pub struct LogResponses<T> {
+    inner: T,
+    level: log::Level,
+    not_ready: bool,
+    target: Option<&'static str>,
+    module_path: Option<&'static str>,
+    file: Option<&'static str>,
+    line: Option<u32>,
+}
+
+
+// ===== impl LogErrors =====
+
+impl<T> LogErrors<T> {
+
+    /// Construct a new `LogErrors` middleware that wraps the given `Service`
+    /// or `NewService`.
+    ///
+    /// The log level will default to `Level::Error` but may be changed with
+    /// the [`at_level`] function.
+    ///
+    /// # Note
+    ///
+    /// Unless the module path of the `LogErrors` middleware is changed with
+    /// the [`in_module`]  methods, log records produced by the returned
+    /// middleware will always have the module path `tower-log-errors`. It may
+    /// be preferred to use the [`log_errors!`] macro instead, as it will
+    /// produce log records which appear to have been produced at the call site.
+    ///
+    /// [`at_level`]: struct.LogErrors.html#method.at_level
+    /// [`in_module`]: struct.LogErrors.html#method.in_module
+    /// [`log_errors!`]: macro.log_errors.html
+    pub fn new(inner: T) -> Self {
+        LogErrors {
+            inner,
+            level: log::Level::Error,
+            target: None,
+            module_path: None,
+            file: None,
+            line: None,
+        }
+    }
+
+    /// Set the log level of the produced log records.
+    ///
+    /// Log records will be logged at the `Error` level by default.
+    pub fn at_level(mut self, level: log::Level) -> Self {
+        self.level = level;
+        self
+    }
+
+    /// Set the target of the produced log records.
+    ///
+    /// The target will default to the module path of the `LogErrors` middleware
+    /// by default.
+    pub fn with_target(mut self, target: &'static str) -> Self {
+        self.target = Some(target);
+        self
+    }
+
+    /// Set the module path of the produced log records to the given string.
+    pub fn in_module(mut self, module_path: &'static str) -> Self {
+        self.module_path = Some(module_path);
+        self
+    }
+
+    /// Set the file and line number of the produced log records.
+    pub fn at_location(mut self, file: &'static str, line: u32) -> Self {
+        self.file = Some(file);
+        self.line = Some(line);
+        self
+    }
+
+    fn child<U>(&self, inner: U) -> LogErrors<U> {
+        LogErrors {
```

Comment:
This struct looks pretty big. I'm wondering if this would work out better storing the config in an `Arc` and cloning. But, this could be figured out later too.